File: sandy/sandy/report/overtime_additional_salary/overtime_additional_salary.py
import frappe
from frappe import _
from frappe.utils import flt


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {
            "fieldname": "name",
            "label": _("Overtime Entry ID"),
            "fieldtype": "Link",
            "options": "Overtime Entry",
            "width": 150
        },
        {
            "fieldname": "employee",
            "label": _("Employee"),
            "fieldtype": "Link",
            "options": "Employee",
            "width": 120
        },
        {
            "fieldname": "employee_name",
            "label": _("Employee Name"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "gender",
            "label": _("Gender"),
            "fieldtype": "Data",
            "width": 80
        },
        {
            "fieldname": "company",
            "label": _("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "width": 120
        },
        {
            "fieldname": "department",
            "label": _("Department"),
            "fieldtype": "Link",
            "options": "Department",
            "width": 120
        },
        {
            "fieldname": "overtime_date",
            "label": _("Overtime Date"),
            "fieldtype": "Date",
            "width": 110
        },
        {
            "fieldname": "overtime_type",
            "label": _("Type"),
            "fieldtype": "Data",
            "width": 90
        },
        {
            "fieldname": "overtime_time",
            "label": _("Time"),
            "fieldtype": "Float",
            "precision": 2,
            "width": 90
        },
        {
            "fieldname": "base_salary",
            "label": _("Base Salary"),
            "fieldtype": "Currency",
            "options": "company:default_currency",
            "width": 120
        },
        {
            "fieldname": "overtime_amount",
            "label": _("Overtime Amount"),
            "fieldtype": "Currency",
            "options": "company:default_currency",
            "width": 140
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "additional_salary_ref",
            "label": _("Additional Salary"),
            "fieldtype": "Link",
            "options": "Additional Salary",
            "width": 150
        }
    ]


def get_data(filters):
    conditions = []

    if filters and filters.get("employee"):
        conditions.append(f"oe.employee = '{frappe.db.escape(filters['employee'])}'")

    if filters and filters.get("department"):
        conditions.append(f"oe.department = '{frappe.db.escape(filters['department'])}'")

    if filters and filters.get("status"):
        conditions.append(f"oe.status = '{frappe.db.escape(filters['status'])}'")

    if filters and filters.get("from_date"):
        conditions.append(f"oe.overtime_date >= '{filters['from_date']}'")

    if filters and filters.get("to_date"):
        conditions.append(f"oe.overtime_date <= '{filters['to_date']}'")

    where_clause = " AND ".join(conditions) if conditions else "1=1"

    data = frappe.db.sql(f"""
        SELECT
            oe.name,
            oe.employee,
            oe.employee_name,
            oe.gender,
            oe.company,
            oe.department,
            oe.overtime_date,
            oe.overtime_type,
            oe.overtime_time,
            oe.base_salary,
            oe.overtime_amount,
            oe.status,
            oe.additional_salary_ref
        FROM
            `tabOvertime Entry` oe
        WHERE
            {where_clause}
        ORDER BY
            oe.overtime_date DESC, oe.creation DESC
    """, as_dict=True)

    return data

File: sandy/sandy/report/overtime_additional_salary/overtime_additional_salary.js
frappe.query_reports["Overtime Additional Salary"] = {
    filters: [
        {
            fieldname: "employee",
            label: __("Employee"),
            fieldtype: "Link",
            options: "Employee"
        },
        {
            fieldname: "department",
            label: __("Department"),
            fieldtype: "Link",
            options: "Department"
        },
        {
            fieldname: "status",
            label: __("Status"),
            fieldtype: "Select",
            options: ["", "Pending", "Salary Created"]
        },
        {
            fieldname: "from_date",
            label: __("From Date"),
            fieldtype: "Date"
        },
        {
            fieldname: "to_date",
            label: __("To Date"),
            fieldtype: "Date"
        }
    ],

    onload(report) {
        report.page.add_inner_button(__("Create Additional Salary"), () => {
            const rows = report.datatable?.getCheckedRows() || [];
            if (!rows.length) {
                frappe.msgprint(__("Please select rows to create Additional Salary"));
                return;
            }

            const selected = rows.map(idx => report.data[idx]);

            frappe.confirm(
                __("Create Additional Salary for {0} selected overtime entries?", [selected.length]),
                () => {
                    frappe.call({
                        method: "sandy.sandy.doctype.overtime_entry.overtime_entry.create_additional_salary",
                        args: {
                            entries: selected.map(d => ({ name: d.name }))
                        },
                        callback(r) {
                            if (r.message && r.message.length) {
                                frappe.msgprint(
                                    __("Created {0} Additional Salary records: {1}",
                                        [r.message.length, r.message.join(", ")]
                                    )
                                );
                                report.refresh();
                            }
                        }
                    });
                }
            );
        });

        frappe.db.get_list("Overtime Entry", {
            filters: { status: "Pending" },
            limit: 1
        }).then(result => {
            if (result.length === 0) {
                frappe.msgprint({
                    title: __("Info"),
                    message: __("No pending overtime entries found."),
                    indicator: "orange"
                });
            }
        });
    },

    formatter(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "status") {
            if (data?.status === "Pending") {
                value = `<span class="indicator-pill red">${value}</span>`;
            } else if (data?.status === "Salary Created") {
                value = `<span class="indicator-pill green">${value}</span>`;
            }
        }

        if (column.fieldname === "overtime_amount" && data?.overtime_amount > 0 && data?.status === "Pending") {
            value = `<strong>${value}</strong>`;
        }

        return value;
    }
};
